Do not export Unchunk, eliminate Shuffle

witchel opened the pull request:
We no longer use Shuffle for testing and only use Unchunk internally to
ArrowInterchange.

This is a rare case of eliminating use in enterprise first, then garbage collecting from open later.

roshandathathri commented:
> It seems like the main thread would always end up accessing data from other NUMA nodes, so binding wouldn't really matter.

No, not necessarily in a parallel section. In a parallel section, every thread can access data from other NUMA nodes due to random accesses but with NUMA blocked for some applications, every thread (including the main thread) can access most of its data on its NUMA node.

> the most common allocation policy for graphs and other large data structures is UseNumaAlloc=false aka allocate interleaved

Yes, but I think this is slightly misleading. For more time-consuming applications like pagerank and mining, NUMA blocked allocation is used because it helps (by ~20% I guess but I'm not sure). 

Note that for distributed execution, we have always used GALOIS_DO_NOT_BIND_THREADS=1 (in our papers) because the dedicated communication thread is not part of the thread pool and binding the threads in the thread pool typically negatively impacts the responsiveness/progress of the communication thread. It is likely hardware dependent too. However, we have not studied this thoroughly in my opinion. For example, we have not tried not-binding the main thread only.

Furthermore note that, for distributed GPU execution where a single machine has multiple GPUs, there are multiple processes on each machine. Even though multiple threads are not used for algorithm execution, they are very useful during graph loading. In this case, we must use GALOIS_DO_NOT_BIND_THREADS=1 for performance (in some cases when it is not used, I have noticed it to hang or take an arbitrarily long time).

ddn0 commented:
> I suggest measuring the performance impact of changing the default behavior on a few different shared-memory and distributed-memory apps before merging the change.

We are due for some performance measurements but I'd like to clean up the state of property graph iterators before embarking (because I know they are sub-optimal). We have an open issue to build some monitoring infrastructure #331. Perhaps we should prioritize that first and concurrently gather experiments we'd like to run (and continue to run) when that's ready?

roshandathathri commented:
> We are due for some performance measurements but I'd like to clean up the state of property graph iterators before embarking (because I know they are sub-optimal). We have an open issue to build some monitoring infrastructure #331. Perhaps we should prioritize that first and concurrently gather experiments we'd like to run (and continue to run) when that's ready?

Ok, that makes sense.

witchel committed:
Do not export Unchunk, eliminate Shuffle

We no longer use Shuffle for testing and only use Unchunk internally to
ArrowInterchange.

tylershunt reviewed (approved):
Seems fine to me.

danielmawhirter commented on libsupport/src/ArrowInterchange.cpp:
```diff
@@ -70,31 +70,33 @@ ApproxArrayDataMemUse(const std::shared_ptr<arrow::ArrayData>& data) {
+  case arrow::StatusCode::IndexError:
+    return ErrorCode::NotFound;
+  default:
+    return ErrorCode::ArrowError;
```
This function was moved to Result.h and should no longer be necessary here

witchel replied:
Thanks for weighing in.  I should have explicitly asked you about this.

What I find confusing is that `ArrowStatusCodeToKatana `is in Result.h, yet files like UpdateGraphView.cpp are calling `ArrowToKatana`.  That was why I left it in, even though it seemed extraneous. When I eliminate `ArrowToKatana`, I get copious build errors.

danielmawhirter replied:
Ok, there might have been concurrent modifications. Uses of `ArrowToKatana` can be replaced with `internal::ArrowStatusCodeToKatana` for now, and should probably eventually migrate to `KATANA_CHECKED`.
Or if you don't want to do that here, that's ok too

danielmawhirter replied:
You meant in the enterprise repo, got it. I'll make that change in https://github.com/KatanaGraph/katana-enterprise/pull/1415
